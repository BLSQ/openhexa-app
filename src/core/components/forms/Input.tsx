import { InputHTMLAttributes, forwardRef, ReactNode } from "react";
import clsx from "clsx";

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  error?: string | null | false | true;
  trailingIcon?: ReactNode;
}

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const {
    type,
    name,
    error,
    className,
    trailingIcon,
    value = "",
    ...delegated
  } = props;

  const inputClassName = clsx(
    "w-full form-input appearance-none relative block",
    "px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm",
    "disabled:opacity-50 disabled:cursor-not-allowed",
    error
      ? "border-red-300  placeholder-red-300  text-red-900  focus:ring-red-500  focus:border-red-500"
      : "border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500",
    trailingIcon && "pr-4",
    className
  );

  return (
    <div className="group relative">
      <input
        id={name}
        name={name}
        type={type ?? "text"}
        aria-invalid={Boolean(error)}
        aria-describedby={`${name}-description`}
        className={inputClassName}
        value={value}
        {...delegated}
        ref={ref}
      />
      {trailingIcon && (
        <div className="absolute inset-y-0 right-0 inline-flex items-center justify-center pr-2.5">
          {trailingIcon}
        </div>
      )}
    </div>
  );
});

Input.displayName = "Input";

export default Input;
