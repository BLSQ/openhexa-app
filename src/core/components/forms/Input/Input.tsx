import { InputHTMLAttributes, forwardRef, ReactNode } from "react";
import clsx from "clsx";

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  error?: string | null | false | true;
  trailingIcon?: ReactNode;
  leading?: ReactNode;
  fullWidth?: boolean;
}

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const {
    type,
    name,
    error,
    defaultValue,
    className,
    trailingIcon,
    leading = null,
    value,
    fullWidth = false,
    ...delegated
  } = props;

  const inputClassName = clsx(
    "form-input appearance-none relative block",
    "px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm",
    "disabled:opacity-50 disabled:cursor-not-allowed",
    error
      ? "border-red-300 placeholder-red-300  text-red-900  focus:ring-red-500  focus:border-red-500"
      : "border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500",
    trailingIcon && "pr-4",
    leading && "pl-10",
    fullWidth && "w-full",
    className,
  );

  return (
    <div className={clsx("group relative", fullWidth && "w-full")}>
      <input
        id={name}
        name={name}
        type={type ?? "text"}
        aria-invalid={Boolean(error)}
        aria-describedby={`${name}-description`}
        className={inputClassName}
        value={defaultValue ? undefined : value ?? ""}
        defaultValue={defaultValue}
        {...delegated}
        ref={ref}
      />
      {leading && (
        <div className="absolute inset-y-0 left-0 z-30 inline-flex items-center justify-center pl-2.5">
          {leading}
        </div>
      )}
      {trailingIcon && (
        <div className="absolute inset-y-0 right-0 z-10 inline-flex items-center justify-center pr-2.5">
          {trailingIcon}
        </div>
      )}
    </div>
  );
});

Input.displayName = "Input";

export default Input;
