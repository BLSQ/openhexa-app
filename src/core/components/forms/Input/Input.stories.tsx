import { Story } from "@ladle/react";
import Input from "./Input";

export const InputStory: Story = (props) => <Input {...props} />;

InputStory.storyName = "default";
InputStory.args = {};
InputStory.argTypes = {};

const defaults = { title: "UI / Forms / Input" };

export default defaults;
