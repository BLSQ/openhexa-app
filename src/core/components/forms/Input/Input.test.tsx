import { EyeIcon } from "@heroicons/react/24/outline";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Input from "../Input";

describe("Input", () => {
  it("renders", async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Input data-testid="input" name="a_field" onChange={onChange} />
    );
    expect(container).toMatchSnapshot();

    const input = screen.getByTestId("input");

    expect(input).not.toBeDisabled();
    expect(input).toHaveValue("");

    await userEvent.type(input, "This is the value");
    expect(onChange).toHaveBeenCalled();
  });

  it("renders a disabled input", async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Input
        disabled
        name="a_field"
        value="Value"
        onChange={onChange}
        data-testid="input"
      />
    );

    const input = screen.getByTestId("input");
    expect(input).toBeInTheDocument();
    expect(input).toBeDisabled();
    await userEvent.type(input, "New value");

    expect(input).toHaveValue("Value");
  });

  it("renders a trailing icon", () => {
    const { container } = render(
      <Input
        value={""}
        onChange={() => {}}
        trailingIcon={<EyeIcon className="w-4" />}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("has an invalid status", () => {
    render(
      <Input error="Err" data-testid="input" value={""} onChange={() => {}} />
    );

    const input = screen.getByTestId("input");

    expect(input.getAttribute("aria-invalid")).toBe("true");
    expect(input).toHaveClass("border-red-300");
  });
});
