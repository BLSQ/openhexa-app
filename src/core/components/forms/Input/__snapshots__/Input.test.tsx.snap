// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders 1`] = `
<div>
  <div
    class="group relative"
  >
    <input
      aria-describedby="a_field-description"
      aria-invalid="false"
      class="form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-hidden focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
      data-testid="input"
      id="a_field"
      name="a_field"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Input renders a trailing icon 1`] = `
<div>
  <div
    class="group relative"
  >
    <input
      aria-describedby="undefined-description"
      aria-invalid="false"
      class="form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-hidden focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500 pr-4"
      type="text"
      value=""
    />
    <div
      class="absolute inset-y-0 right-0 z-10 inline-flex items-center justify-center pr-2.5"
    >
      <svg
        aria-hidden="true"
        class="w-4"
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.036 12.322a1.012 1.012 0 0 1 0-.639C3.423 7.51 7.36 4.5 12 4.5c4.638 0 8.573 3.007 9.963 7.178.07.207.07.431 0 .639C20.577 16.49 16.64 19.5 12 19.5c-4.638 0-8.573-3.007-9.963-7.178Z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
        <path
          d="M15 12a3 3 0 1 1-6 0 3 3 0 0 1 6 0Z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
  </div>
</div>
`;
